automated install/basic-install.sh

```diff
@@ -1990,6 +1990,100 @@ accountForRefactor() {
+    if [[ "${INSTALL_WEB_INTERFACE}" == false ]] && [[ "${INSTALL_WEB_SERVER}" == true ]]; then
+        echo "You can install the Web interface without the web server, but you cannot install the web server without the web interface."
+        echo "INSTALL_WEB_SERVER is set to ${INSTALL_WEB_SERVER} and INSTALL_WEB_INTERFACE is set to ${INSTALL_WEB_INTERFACE}."
+        exit 1
+    fi
+    # Regex patterns to match
+    addressMask="^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)(\/([1-9]|[1-2][0-9]|3[0-2]))$"
+    addressReverse="^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}in-addr\.arpa$"
+    # Validate the IPv4 address and subnet mask
+    if [[ ! ${IPV4_ADDRESS} =~ ${addressMask} ]] ; then
+        echo "${IPV4_ADDRESS} is not a valid address or it is missing the subnet mask"
+        exit 1 
+    else
+        echo "IPv4 ${IPV4_ADDRESS}"
+    fi
+    # Validate the DNS_1 IP addresses
+    if ! valid_ip "${PIHOLE_DNS_1}" ; then
+        echo "${PIHOLE_DNS_1} is not a valid address"
+        exit 1
+    else
+        echo "DNS1 ${PIHOLE_DNS_1}"
+    fi
+    # Validate the DNS_2 IP addresses
+    if ! valid_ip "${PIHOLE_DNS_2}" ; then
+        echo "${PIHOLE_DNS_2} is not a valid address"
+        exit 1
+    else
+        echo "DNS2 ${PIHOLE_DNS_2}"
+    fi
+    # Validate Web UI Layout
+    if [[ "${WEBUIBOXEDLAYOUT}" == "boxed" ]] || [[ "${WEBUIBOXEDLAYOUT}" == "traditional" ]]; then
+            echo "WEBUIBOXEDLAYOUT is ${WEBUIBOXEDLAYOUT}"
+        elif [[ -z "${WEBUIBOXEDLAYOUT}" ]]; then
+            echo "WEBUIBOXEDLAYOUT is not set"
+        else
+            echo "WEBUIBOXEDLAYOUT valid options are: boxed or traditional."
+            exit 1
+        fi
+    # Validate Conditional Forwarding settings
+    if ! valid_ip "${CONDITIONAL_FORWARDING_IP}" ; then
+            echo "CONDITIONAL_FORWARDING_IP is ${CONDITIONAL_FORWARDING_IP}"
+    elif [[ -z "${CONDITIONAL_FORWARDING_IP}" ]]; then
+            echo "CONDITIONAL_FORWARDING_IP is not set"
+    else
+            echo "${CONDITIONAL_FORWARDING_IP} is not a valid IP Address"
+            exit 1
+        fi
+    if [[ "${CONDITIONAL_FORWARDING_REVERSE}" =~ ${addressReverse} ]]; then
+            echo "CONDITIONAL_FORWARDING_REVERSE is ${CONDITIONAL_FORWARDING_REVERSE}"
+    elif [[ -z "${CONDITIONAL_FORWARDING_REVERSE}" ]]; then
+            echo "CONDITIONAL_FORWARDING_REVERSE is not set"
+    else
+            echo "${CONDITIONAL_FORWARDING_REVERSE} is not a valid reverse address"
+            exit 1
+    fi
+    # LIGHTTPD_ENABLED should not be written to the setupVars.conf file.  It's a temporary variable and not used for installation or updates.
+    sed -i '/LIGHTTPD_ENABLED=/ d' "${setupVars}"
+    # DNSMASQ_LISTENING is not used during setup or updates.
+    sed -i '/DNSMASQ_LISTENING=/ d' "${setupVars}"
+    # Ensure setupVars ends with a newline
+    sed -i -e '$a\ ' "${setupVars}"
```

Comment:
If I am not mistaken, since the file is never fully removed or overwritten, this would add an additional newline on every update/reinstall. AFAIK bash has no problem to source files with missing trailing newline, does it?

Reply:
It will make sure there is exactly one blank line at the end of the file.  Off the top of my head, I don't remember if it was bash that had the problem or php (because the same file is used for both).  I do remember it will fail if the blank line is not there.

Reply:
As a side note, I've have now enabled the options in VS Code to automatically remove trailing white spaces and ensure each file ends with exactly one blank line.

Reply:
> It will make sure there is exactly one blank line at the end of the file.

That is not true, the sed will always add an additional newline, regardless if there is already one. Just tested in bash shell to be sure, but the syntax is clear on it's own. It would require a test for an existing character.
Ah it even adds a space instead of an empty like.

An empty new line would be `sed -i -e '$a\\n' "${setupVars}"` or better `echo >> "${setupVars}"`. I'm currently failing on the syntax for "if last line an it contains a character" with sed 😄.

Reply:
There it is:
```
sed -i '$s/^..*$/\0\n/' "${setupVars}"
```
or
```
sed -Ei '$s/^.+$/\0\n/' "${setupVars}"
```
but I like basic regex where possible.